pkg/cri/server/sandbox_stats_windows.go

```diff
@@ -19,17 +19,397 @@ package server
+	if err != nil {
+		return nil, err
+	}
+
+	podSandboxStats.Windows.Process = &runtime.WindowsProcessUsage{
+		Timestamp:    timestamp.UnixNano(),
+		ProcessCount: &runtime.UInt64Value{Value: pidCount},
+	}
+
+	c.saveSandBoxMetrics(podSandboxStats.Attributes.Id, podSandboxStats)
+
+	return podSandboxStats, nil
+}
+
+func convertMetricsToWindowsStats(metrics []*types.Metric) (map[string]*wstats.Statistics, error) {
+	statsMap := make(map[string]*wstats.Statistics)
+	for _, stat := range metrics {
+		var podRuntimeStats *wstats.Statistics
+		podStatsData, err := typeurl.UnmarshalAny(stat.Data)
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract container metrics: %w", err)
+		}
+
+		// extract the metrics if available
+		// podStatsData can be nil for pods that don't have an actual podsandbox
+		// this is the case for hostprocess pods
+		if podStats, ok := podStatsData.(*wstats.Statistics); ok {
+			podRuntimeStats = podStats
+		}
+
+		statsMap[stat.ID] = podRuntimeStats
+	}
+	return statsMap, nil
+}
+
+func (c *criService) toPodSandboxStats(sandbox sandboxstore.Sandbox, statsMap map[string]*wstats.Statistics, containers []containerstore.Container, timestamp time.Time) (*runtime.WindowsContainerStats, []*runtime.WindowsContainerStats, error) {
+	podMetric, ok := statsMap[sandbox.ID]
+	if !ok {
+		return nil, nil, fmt.Errorf("failed to extract container metrics: no pod stats found")
+	}
+
+	podRuntimeStats, err := c.windowsCPUAndMemoryStats(podMetric)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to extract container metrics: %w", err)
+	}
+
+	windowsContainerStats := []*runtime.WindowsContainerStats{}
+	for _, cntr := range containers {
+		containerMetric := statsMap[cntr.ID]
+
+		if containerMetric == nil {
+			// maybe just a warning?
+			return nil, nil, fmt.Errorf("failed find metric ")
+		}
+
+		containerStats, err := c.windowsCPUAndMemoryStats(containerMetric)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to extract container metrics: %w", err)
+		}
+
+		// Calculate NanoCores for container
+		if containerStats.Cpu.UsageCoreNanoSeconds != nil {
+			nanoCoreUsage := getUsageNanoCores(containerStats.Cpu.UsageCoreNanoSeconds.Value, cntr.Stats, containerStats.Cpu.Timestamp)
+			containerStats.Cpu.UsageNanoCores = &runtime.UInt64Value{Value: nanoCoreUsage}
+		}
+
+		// On Windows we need to add up all the podStatsData to get the Total for the Pod as there isn't something
+		// like a parent cgroup that queried for all the pod podStatsData
+		appendCPUPodStats(podRuntimeStats, containerStats, timestamp)
+		appendMemoryPodStats(podRuntimeStats, containerStats, timestamp)
+
+		// If snapshotstore doesn't have cached snapshot information
+		// set WritableLayer usage to zero
+		var usedBytes uint64
+		sn, err := c.snapshotStore.Get(cntr.ID)
+		if err == nil {
+			usedBytes = sn.Size
+		}
+		containerStats.WritableLayer = &runtime.WindowsFilesystemUsage{
+			Timestamp: sn.Timestamp,
```

Comment:
Is it safe to use `sn.Timestamp` even if `err != nil` above?

Reply:
makes sense that it might not be but seems to not be an issue?  I've copied this section from 

https://github.com/containerd/containerd/blob/0d6881898e3635eee7b280211a0d22f386c82728/pkg/cri/server/container_stats_list_linux.go#L39-L47

and it is also here: https://github.com/containerd/containerd/blob/bbe46b8c43fc2febe316775bc2d4b9d697bbf05c/pkg/cri/server/container_stats_list_windows.go#L38-L45

Reply:
Guess it's okay then :)